airbyte-integrations/connectors/source-surveymonkey/source_surveymonkey/streams.py

```diff
@@ -0,0 +1,257 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+import tempfile
+import urllib.parse
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional
+
+import pendulum
+import requests
+import vcr
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams.http import HttpStream
+
+cache_file = tempfile.NamedTemporaryFile()
+
+
+class SurveymonkeyStream(HttpStream, ABC):
+    url_base = "https://api.surveymonkey.com/v3/"
+    primary_key = "id"
+    data_field = "data"
+
+    def __init__(self, start_date: pendulum.datetime, **kwargs):
+        self._start_date = start_date
+        super().__init__(**kwargs)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        resp_json = response.json()
+        links = resp_json.get("links", {})
+        if links.get("next"):
+            next_query_string = urllib.parse.urlsplit(links["next"]).query
+            params = dict(urllib.parse.parse_qsl(next_query_string))
+            return params
+
+    def request_headers(self, **kwargs) -> Mapping[str, Any]:
+        return {"Content-Type": "application/json"}
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        if response_json.get("error"):
+            raise Exception(repr(response_json.get("error")))
+        result = response_json.get(self.data_field, [])
+        yield from result
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return next_page_token or {}
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        with vcr.use_cassette(cache_file.name, record_mode="new_episodes", serializer="json"):
```

Comment:
can you add a comment describing why: 
* we are using vcr
* why we use the `new_episodes` mode. The description says: "Record new interactions. Replay previously recorded interactions. It is similar to the once record mode, but will always record new interactions, even if you have an existing recorded one that is similar, but not identical.". Is the idea that we just use it to avoid throwing an error if there is a new request? 


also shouldn't this functionality be added only to the surveys stream, not every stream?

Reply:
I want to cache *all* requests to all endpoints during iteration.  This API is very very limited, we need to reuse everything possible. "Cause an error to be raised for new requests if there is a cassette file" is not that I need. I need to save all of them and reuse all of them. In stream slices, in details, etc.

Reply:
@vovavovavovavova this comment with explanation should be in the source code as well